apps/designer/app/designer/features/sidebar-left/panels/pages/settings.tsx

```diff
@@ -7,26 +7,119 @@ import {
   useId,
   styled,
   Flex,
-  toast,
+  toastNonFieldErrors,
   InputErrorsTooltip,
 } from "@webstudio-is/design-system";
-import { useFetcher } from "@remix-run/react";
-import { ChevronDoubleLeftIcon } from "@webstudio-is/icons";
+import { useFetcher, type Fetcher } from "@remix-run/react";
+import { ChevronDoubleLeftIcon, TrashIcon } from "@webstudio-is/icons";
+import { utils as projectUtils } from "@webstudio-is/project";
 import type { ZodError } from "zod";
 import { Header } from "../../lib/header";
-import { useEffect, useRef, useState } from "react";
+import { useRef, useState, useCallback } from "react";
 import { type Page } from "@webstudio-is/project";
+import { usePages } from "~/designer/shared/nano-states";
+import { useDebounce, useUnmount } from "react-use";
+import { useOnFetchEnd, usePersistentFetcher } from "~/shared/fetcher";
 
 const Group = styled(Flex, {
   marginBottom: "$3",
   gap: "$2",
   defaultVariants: { direction: "column" },
 });
 
-export type CreatePageData =
-  | { ok: true; page: Page }
+type FetcherData<Payload> =
+  | ({ ok: true } & Payload)
   | { errors: string | ZodError["formErrors"] };
 
+const normalizeErrors = (
+  errors: string | ZodError["formErrors"]
+): ZodError["formErrors"] =>
+  typeof errors === "string"
+    ? { formErrors: [errors], fieldErrors: {} }
+    : errors;
+
+type EditablePage = Omit<Page, "treeId" | "id">;
+
+const FormFields = ({
+  disabled,
+  values,
+  onChange,
+  fetcher,
+}: {
+  disabled?: boolean;
+  values?: EditablePage;
+  onChange?: <FieldName extends keyof EditablePage>(
+    field: FieldName,
+    value: EditablePage[FieldName]
+  ) => void;
+  fetcher: Fetcher<FetcherData<unknown>>;
+}) => {
+  const fieldIds = {
+    name: useId(),
+    path: useId(),
+  } as const;
+
+  const fieldNames = useRef(Object.keys(fieldIds));
+
+  const [fieldErrors, setFieldErrors] = useState<
+    ZodError["formErrors"]["fieldErrors"]
+  >({});
+
+  useOnFetchEnd(fetcher, (data) => {
+    if ("errors" in data) {
+      const errors = normalizeErrors(data.errors);
+      toastNonFieldErrors(errors, fieldNames.current);
+      setFieldErrors(errors.fieldErrors);
+    }
+  });
+
+  const resetErrors = (fieldName: keyof typeof fieldIds) =>
+    setFieldErrors(({ [fieldName]: _, ...rest }) => rest);
+
+  return (
+    <>
+      <Group>
+        <Label htmlFor={fieldIds.name} size={2}>
+          Page Name
+        </Label>
+        <InputErrorsTooltip errors={fieldErrors.name}>
+          <TextField
+            state={fieldErrors.name && "invalid"}
+            id={fieldIds.name}
+            name="name"
+            disabled={disabled}
+            value={values?.name}
+            onChange={(event) => {
+              resetErrors("name");
+              onChange?.("name", event.target.value);
```

Comment:
its better to have a single argument in in those handlers, like with the native event handlers, in this case:
`onChange({field: 'name', value: event.target.value})`

On the related note: is it really necessary to handle onChange for every field separately instead of handling onChange on the form component and then, new FormData(event.currentTarget) and then validate the entire form data every time, so you have a single place for entire form

Reply:
There're couple reasons handle changes separately:

1. We want to send to the server only changed data because it reduces the risk of overriding more recent data on the server with potentially outdated data on the client, also a bit smaller payloads.
2. It's easier to track whether we have data to submit when we have a separate object with only changes.

Reply:
I think you are doing it here so you can know which field has changed and submit only changed fields. Another way of doing it which is potentially simpler: just iterate over the data and compare the fields before submission

Reply:
I think the current way is simpler. I'd keep it as is in this PR, and consider again when I will work on adding "meta" support (nested repeating forms), because it may require refactoring anyway.

Reply:
current way spreads the logic for how you handle this data all over the form, it looks simpler now but its more custom and more code

finding changed fields is actually quite easy, its a simple loop you would do once before validation and submission, it will allow you to get rid of all the logic in the form and make it headless, then handle the validation/submission in a separate hook from outside

Reply:
> finding changed fields is actually quite easy, its a simple loop you would do once before validation and submission,

We need both representations anyway complete and partial. But it's easier to go from partial to complete (`latest = {...onServer, ...changes}`) than to go from complete to partial (`changes = diff(onServer, latest)`)

> it will allow you to get rid of all the logic in the form and make it headless, then handle the validation/submission in a separate hook from outside

This is separate issue, you've mentioned below. Will extract to a hook, yes.